File: sdk/go/webhooks_test.go
package omniswitch

import (
	"fmt"
	"strings"
	"testing"
	"time"
)

const testSecret64 = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
const testBody = `{"event":"dispute.created","paymentId":"pay_1"}`

func signWebhook(secret, body string, timestamp int64) string {
	signature := hmacSHA256Hex(secret, fmt.Sprintf("%d.%s", timestamp, body))
	return fmt.Sprintf("t=%d,v1=%s", timestamp, signature)
}

func TestAcceptsACorrectlySignedFreshPayload(t *testing.T) {
	header := signWebhook(testSecret64, testBody, time.Now().Unix())
	if !VerifyWebhookSignature(testSecret64, testBody, header) {
		t.Error("expected true")
	}
}

func TestRejectsAPayloadSignedWithTheWrongSecret(t *testing.T) {
	header := signWebhook(strings.Repeat("b", 64), testBody, time.Now().Unix())
	if VerifyWebhookSignature(testSecret64, testBody, header) {
		t.Error("expected false")
	}
}

func TestRejectsAMutatedBodyAgainstASignatureComputedForTheOriginal(t *testing.T) {
	header := signWebhook(testSecret64, testBody, time.Now().Unix())
	if VerifyWebhookSignature(testSecret64, testBody+"tampered", header) {
		t.Error("expected false")
	}
}

func TestRejectsAMissingSignatureHeader(t *testing.T) {
	if VerifyWebhookSignature(testSecret64, testBody, "") {
		t.Error("expected false for empty header")
	}
}

func TestRejectsAMalformedHeader(t *testing.T) {
	cases := []string{"v1=deadbeef", "t=1700000000", "garbage"}
	for _, c := range cases {
		if VerifyWebhookSignature(testSecret64, testBody, c) {
			t.Errorf("expected false for header %q", c)
		}
	}
}

func TestRejectsATimestampOutsideTheToleranceWindow(t *testing.T) {
	staleTimestamp := time.Now().Unix() - 10*60
	header := signWebhook(testSecret64, testBody, staleTimestamp)
	if VerifyWebhookSignature(testSecret64, testBody, header) {
		t.Error("expected false")
	}
}

func TestAcceptsACustomToleranceWindow(t *testing.T) {
	timestamp := time.Now().Unix() - 60
	header := signWebhook(testSecret64, testBody, timestamp)
	if VerifyWebhookSignatureWithTolerance(testSecret64, testBody, header, 30) {
		t.Error("expected false with 30s tolerance")
	}
	if !VerifyWebhookSignatureWithTolerance(testSecret64, testBody, header, 120) {
		t.Error("expected true with 120s tolerance")
	}
}

func TestRejectsANonHexV1ValueWithoutPanicking(t *testing.T) {
	header := "t=1700000000,v1=not-hex!!"
	if VerifyWebhookSignature(testSecret64, testBody, header) {
		t.Error("expected false")
	}
}

File: sdk/java/src/test/java/io/omniswitch/sdk/WebhooksTest.java
package io.omniswitch.sdk;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;

import org.junit.jupiter.api.Test;

class WebhooksTest {

  private static final String SECRET = "a".repeat(64);
  private static final String BODY = "{\"event\":\"dispute.created\",\"paymentId\":\"pay_1\"}";

  private static String sign(String secret, String body) {
    return sign(secret, body, System.currentTimeMillis() / 1000);
  }

  private static String sign(String secret, String body, long timestamp) {
    String signature = Signing.hmacSha256Hex(secret, timestamp + "." + body);
    return "t=" + timestamp + ",v1=" + signature;
  }

  @Test
  void acceptsACorrectlySignedFreshPayload() {
    assertTrue(Webhooks.verifyWebhookSignature(SECRET, BODY, sign(SECRET, BODY)));
  }

  @Test
  void rejectsAPayloadSignedWithTheWrongSecret() {
    assertFalse(Webhooks.verifyWebhookSignature(SECRET, BODY, sign("b".repeat(64), BODY)));
  }

  @Test
  void rejectsAMutatedBodyAgainstASignatureComputedForTheOriginal() {
    String header = sign(SECRET, BODY);
    assertFalse(Webhooks.verifyWebhookSignature(SECRET, BODY + "tampered", header));
  }

  @Test
  void rejectsAMissingSignatureHeader() {
    assertFalse(Webhooks.verifyWebhookSignature(SECRET, BODY, null));
    assertFalse(Webhooks.verifyWebhookSignature(SECRET, BODY, ""));
  }

  @Test
  void rejectsAMalformedHeader() {
    assertFalse(Webhooks.verifyWebhookSignature(SECRET, BODY, "v1=deadbeef"));
    assertFalse(Webhooks.verifyWebhookSignature(SECRET, BODY, "t=1700000000"));
    assertFalse(Webhooks.verifyWebhookSignature(SECRET, BODY, "garbage"));
  }

  @Test
  void rejectsATimestampOutsideTheToleranceWindow() {
    long staleTimestamp = System.currentTimeMillis() / 1000 - 10 * 60;
    String header = sign(SECRET, BODY, staleTimestamp);
    assertFalse(Webhooks.verifyWebhookSignature(SECRET, BODY, header));
  }

  @Test
  void acceptsACustomToleranceWindow() {
    long timestamp = System.currentTimeMillis() / 1000 - 60;
    String header = sign(SECRET, BODY, timestamp);
    assertFalse(Webhooks.verifyWebhookSignature(SECRET, BODY, header, 30));
    assertTrue(Webhooks.verifyWebhookSignature(SECRET, BODY, header, 120));
  }

  @Test
  void rejectsANonHexV1ValueWithoutThrowing() {
    String header = "t=1700000000,v1=not-hex!!";
    assertDoesNotThrow(() -> Webhooks.verifyWebhookSignature(SECRET, BODY, header));
    assertFalse(Webhooks.verifyWebhookSignature(SECRET, BODY, header));
  }
}
